File: mizhara-backend/lib/tracking.go
package lib

import (
	"fmt"
	"net/url"
	"strings"
)

type TrackingProvider string

const (
	ProviderDelhivery  TrackingProvider = "delhivery"
	ProviderBlueDart   TrackingProvider = "bluedart"
	ProviderDTDC       TrackingProvider = "dtdc"
	ProviderIndiaPost  TrackingProvider = "indiapost"
	ProviderShiprocket TrackingProvider = "shiprocket"
	ProviderOther      TrackingProvider = "other"
)

func BuildTrackingURL(provider TrackingProvider, trackingNumber, customURL string) string {
	trimmed := strings.TrimSpace(trackingNumber)
	if trimmed == "" {
		return ""
	}
	if strings.TrimSpace(customURL) != "" {
		return strings.TrimSpace(customURL)
	}
	if provider == ProviderOther {
		return ""
	}
	enc := url.QueryEscape(trimmed)
	switch provider {
	case ProviderDelhivery:
		return fmt.Sprintf("https://www.delhivery.com/track/package/%s", enc)
	case ProviderBlueDart:
		return fmt.Sprintf("https://www.bluedart.com/web/guest/trackdartresultthirdparty?trackFor=0&trackNo=%s", enc)
	case ProviderDTDC:
		return fmt.Sprintf("https://www.dtdc.in/tracking.asp?strCnno=%s", enc)
	case ProviderIndiaPost:
		return fmt.Sprintf("https://www.indiapost.gov.in/_layouts/15/DOP.Portal.Tracking/TrackConsignment.aspx?consignmentnumber=%s", enc)
	case ProviderShiprocket:
		return fmt.Sprintf("https://shiprocket.co/tracking/%s", enc)
	default:
		return ""
	}
}

File: mizhara-web/src/constants/tracking.ts
import type { TrackingProvider } from "@/types/order";

export const DEFAULT_TRACKING_PROVIDER: TrackingProvider = "delhivery";

export const TRACKING_URLS: Record<Exclude<TrackingProvider, "other">, (trackingNumber: string) => string> = {
  delhivery: (n) => `https://www.delhivery.com/track/package/${encodeURIComponent(n)}`,
  bluedart: (n) =>
    `https://www.bluedart.com/web/guest/trackdartresultthirdparty?trackFor=0&trackNo=${encodeURIComponent(n)}`,
  dtdc: (n) => `https://www.dtdc.in/tracking.asp?strCnno=${encodeURIComponent(n)}`,
  indiapost: (n) =>
    `https://www.indiapost.gov.in/_layouts/15/DOP.Portal.Tracking/TrackConsignment.aspx?consignmentnumber=${encodeURIComponent(n)}`,
  shiprocket: (n) => `https://shiprocket.co/tracking/${encodeURIComponent(n)}`,
};

export const TRACKING_PROVIDERS: { id: TrackingProvider; label: string }[] = [
  { id: "delhivery", label: "Delhivery" },
  { id: "bluedart", label: "Blue Dart" },
  { id: "dtdc", label: "DTDC" },
  { id: "indiapost", label: "India Post" },
  { id: "shiprocket", label: "Shiprocket" },
  { id: "other", label: "Other" },
];
